File: backend/app/domain/athlete_profile.py
from datetime import date

from pydantic import AwareDatetime, BaseModel, Field

from app.domain.activities import ActivityKind


class ActivityAnalysisRecord(BaseModel):
    """Provider-neutral historical activity facts used by athlete analysis."""

    activity_kind: ActivityKind | None
    started_at: AwareDatetime
    distance_meters: float = Field(ge=0)
    moving_time_seconds: int = Field(ge=0)
    elevation_gain_meters: float | None = Field(default=None, ge=0)


class RepresentativeRange(BaseModel):
    """Percentile distribution for one metric with its valid sample count."""

    sample_size: int = Field(ge=1)
    p25: float
    median: float
    p75: float
    p90: float


class ActivityCapabilityRanges(BaseModel):
    """Typed representative per-activity ranges for one activity kind."""

    distance_meters: RepresentativeRange
    moving_time_seconds: RepresentativeRange
    elevation_gain_meters: RepresentativeRange | None = None
    elevation_gain_meters_per_km: RepresentativeRange | None = None
    pace_seconds_per_km: RepresentativeRange | None = None
    average_moving_speed_meters_per_second: RepresentativeRange | None = None


class ActivityKindSummary(BaseModel):
    """Deterministic aggregates for one represented RouteMuse activity kind."""

    activity_kind: ActivityKind
    activity_count: int = Field(ge=1)
    total_distance_meters: float = Field(ge=0)
    total_moving_time_seconds: int = Field(ge=0)
    total_elevation_gain_meters: float | None = Field(default=None, ge=0)
    elevation_sample_count: int = Field(ge=0)
    active_weeks: int = Field(ge=1)
    median_distance_meters: float = Field(ge=0)
    median_moving_time_seconds: float = Field(ge=0)
    median_elevation_gain_meters: float | None = Field(default=None, ge=0)
    capability_ranges: ActivityCapabilityRanges


class DominantActivityResult(BaseModel):
    """The selected activity kind and the facts supporting that selection."""

    activity_kind: ActivityKind
    total_moving_time_seconds: int = Field(ge=0)
    activity_count: int = Field(ge=1)
    total_distance_meters: float = Field(ge=0)
    moving_time_share: float = Field(ge=0, le=1)


class AthleteProfile(BaseModel):
    period_start: date
    period_end: date
    timezone: str
    activities_analyzed: int = Field(default=0, ge=0)
    unsupported_activities_excluded: int = Field(default=0, ge=0)
    activity_summaries: list[ActivityKindSummary] = Field(default_factory=list)
    dominant_activity: DominantActivityResult | None = None

File: frontend/types/athleteProfile.ts
export type ActivityKind =
  | "walking"
  | "running"
  | "trail_running"
  | "hiking"
  | "road_cycling"
  | "gravel_cycling"
  | "mountain_biking"
  | "alpine_skiing"
  | "backcountry_skiing"
  | "nordic_skiing";

export type AthleteProfileRequest = {
  start_date: string;
  end_date: string;
  timezone: string;
};

export type RepresentativeRange = {
  sample_size: number;
  p25: number;
  median: number;
  p75: number;
  p90: number;
};

export type ActivityCapabilityRanges = {
  distance_meters: RepresentativeRange;
  moving_time_seconds: RepresentativeRange;
  elevation_gain_meters: RepresentativeRange | null;
  elevation_gain_meters_per_km: RepresentativeRange | null;
  pace_seconds_per_km: RepresentativeRange | null;
  average_moving_speed_meters_per_second: RepresentativeRange | null;
};

export type ActivityKindSummary = {
  activity_kind: ActivityKind;
  activity_count: number;
  total_distance_meters: number;
  total_moving_time_seconds: number;
  total_elevation_gain_meters: number | null;
  elevation_sample_count: number;
  active_weeks: number;
  median_distance_meters: number;
  median_moving_time_seconds: number;
  median_elevation_gain_meters: number | null;
  capability_ranges: ActivityCapabilityRanges;
};

export type DominantActivity = {
  activity_kind: ActivityKind;
  total_moving_time_seconds: number;
  activity_count: number;
  total_distance_meters: number;
  moving_time_share: number;
};

export type ActivityVolume = {
  activity_count: number;
  moving_time_seconds: number;
  distance_meters: number;
  active_weeks: number;
};

export type WeeklyActivityVolume = {
  activities_per_week: number;
  moving_time_seconds_per_week: number;
  distance_meters_per_week: number;
};

export type HistoricalBaselineSignals = {
  period_start: string;
  period_end: string;
  effective_days: number;
  volume: ActivityVolume;
  weekly_volume: WeeklyActivityVolume;
};

export type RecentToBaselineRatios = {
  activities_per_week_ratio: number | null;
  moving_time_seconds_per_week_ratio: number | null;
  distance_meters_per_week_ratio: number | null;
};

export type RecencySignals = {
  nominal_window_days: number;
  effective_window_days: number;
  window_start: string;
  window_end: string;
  volume: ActivityVolume;
  weekly_volume: WeeklyActivityVolume;
  baseline: HistoricalBaselineSignals | null;
  recent_to_baseline: RecentToBaselineRatios | null;
};

export type ConsistencySignals = {
  activity_kind: ActivityKind;
  calendar_weeks: number;
  active_week_ratio: number;
  activities_per_week: number;
  longest_inactivity_gap_days: number;
  days_since_last_activity: number;
  recency: RecencySignals;
};

export type AthleteProfile = {
  period_start: string;
  period_end: string;
  timezone: string;
  activities_analyzed: number;
  unsupported_activities_excluded: number;
  activity_summaries: ActivityKindSummary[];
  dominant_activity: DominantActivity | null;
  consistency_signals: ConsistencySignals[];
};
